src/util/blockfilter.rs

```diff
@@ -0,0 +1,626 @@
+// Written in 2019 by
+//   The rust-bitcoin developers
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+// This module was largely copied from https://github.com/rust-bitcoin/murmel/blob/master/src/blockfilter.rs
+// on 11. June 2019 which is licensed under Apache, that file specifically
+// was written entirely by Tamas Blummer, who is re-licensing its contents here as CC0.
+
+//!
+//! # BIP158 Compact Block Filters for Light Clients
+//!
+//! Implements a structure for compact filters on block data, for use in the BIP 157 light client protocol.
+//! The filter construction proposed is an alternative to Bloom filters, as used in BIP 37,
+//! that minimizes filter size by using Golomb-Rice coding for compression.
+//!
+
+use blockdata::{
+        transaction::OutPoint,
+        block::Block,
+        script::Script
+    };
+use consensus::{Decodable, Encodable, encode::VarInt};
+use util::hash::BitcoinHash;
+use bitcoin_hashes::sha256d::Hash as Sha256dHash;
+use bitcoin_hashes::Hash;
+use byteorder::{ByteOrder, LittleEndian};
+use siphasher::sip::SipHasher;
+use std::{
+    cmp,
+    collections::HashSet,
+    hash::Hasher,
+    io::{self, Cursor},
+    error::Error,
+    fmt,
+    fmt::{Debug, Display, Formatter}
+};
+
+/// BIP158 base filter type 0: input and output scripts
+pub const SCRIPT_FILTER:u8 = 0;
+
+const P: u8 = 19;
+const M: u64 = 784931;
+
+/// Error raised if script for a spent UTXO can not be retrieved
+pub struct ScriptAccessError {
+    coin: OutPoint
+}
+
+impl Error for ScriptAccessError {
+    fn description(&self) -> &str {
+        "unresolved UTXO"
+    }
+}
+
+impl Display for ScriptAccessError {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
+        write!(f, "unresolved UTXO {}", self.coin)
+    }
+}
+
+impl Debug for ScriptAccessError {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
+        write!(f, "unresolved UTXO {}", self.coin)
+    }
+}
+
+
+/// interface to an object capable of retrieving a spent script
+pub trait ScriptAccessor {
+    /// get script for a spent outpoint
+    fn get_scripts(&self, coins: Vec<OutPoint>) -> Result<Vec<Script>, Box<Error>>;
+}
+
+/// a computed or read block filter
+pub struct BlockFilter {
+    /// id of the block
+    pub block: Sha256dHash,
+    /// filte type (see SCRIPT_FILTER)
+    pub filter_type: u8,
+    /// Golomb encoded filter
+    pub content: Vec<u8>
+}
+
+impl BlockFilter {
+    /// Compute a SCRIPT_FILTER that contains spent and output scripts
+    pub fn compute_script_filter(block: &Block, scripts: Box<ScriptAccessor>) -> Result<BlockFilter, Box<Error>> {
```

Comment:
I do not like non-trivial closures, there are ugly examples for that pattern eg. in tokio. The Box<dyn Trait> is less cumbersome and the call overhead does not matter as it is called once to resolve all coins referred, which is likely optimal for a db storing utxos.

Reply:
Could you at least parameterize over `ScriptAccessor` instead of using a `Box`? Every allocation is an `std` dependency and contributes to memory fragmentation and requires users to type `Box::new`.

Reply:
I can make this a closure if you prefer. I understand implications of using Box, but is it really an issue here given that the retrieved objects are Scripts from TxOuts that were defined by you as Box<[u8]> ?

Reply:
I think the API is fine. Added use comments in module header.

Reply:
If you just do it like this:
```rust
     pub fn compute_script_filter<A: ScriptAccessor>(block: &Block, scripts: A) -> Result<BlockFilter, Box<Error>>
```
Then one can implement `impl ScriptAccessor for Box<Whatever>`, right? Or f.e. `impl<'a> ScriptAccessor for &'a Whatever`. Then the decision to heap-allocate is on the user.

Reply:
parametrized